packages/chat-stateful-client/src/ChatContext.ts

```diff
@@ -50,21 +50,15 @@ export class ChatContext {
     return this._state;
   }
 
-  public setState(state: ChatClientState): void {
-    this._state = state;
-    if (!this._batchMode) {
-      this._emitter.emit('stateChanged', this._state);
-    }
-  }
-
   public modifyState(modifier: (draft: ChatClientState) => void): void {
+    const priorState = this._state;
     this._state = produce(this._state, modifier, (patches: Patch[]) => {
       if (getLogLevel() === 'verbose') {
         // Log to `info` because AzureLogger.verbose() doesn't show up in console.
         this._logger.info(`State change: ${JSON.stringify(patches)}`);
       }
     });
-    if (!this._batchMode) {
+    if (!this._batchMode && this._state !== priorState) {
```

Comment:
Does it handle complex deep changes?
My knowledge around javascript object comparison is rusty.
But what happens if a deeply nested attribute has changed in the state. Would this comparison handle that as well?

Reply:
Approved. However, I would appreciate a comment on this. :)

Reply:
Nope. This is just a reference check.

The magic part is that the way we modify state is on line 55 via `produce`. *That* ensures that if there any change in the dag, the root node is recreated.

This is why selectors are effecient and React can skip renders when there is no change.